clustering/ejb/WFLY-7628_Distributed_EJB_Timers.adoc

```diff
@@ -0,0 +1,142 @@
+// Discuss with QE during the Kickoff state to decide this
+* [ ] Engineering
+
+* [x] QE
+
+=== Affected Projects or Components
+
+=== Other Interested Projects
+
+=== Relevant Installation Types
+// Remove the x next to the relevant field if the feature in question is not relevant
+// to that kind of WildFly installation
+* [x] Traditional standalone server (unzipped or provisioned by Galleon)
+
+* [x] Managed domain
+
+* [x] OpenShift s2i
+
+* [x] Bootable jar
+
+== Requirements
+
+=== Hard Requirements
+
+* Introduce a proper SPI for a TimerService within org.jboss.as.ejb3.timerservice.spi, to facilitate multiple implementations.
+** Adapt the existing TimerResource runtime attributes/operations to use the TimerService SPI instead of referencing specific implementation classes.
+** Adapt existing TimerService implementation to this SPI.
+* Add a CompositeTimerService implementation that delegates to separate persistent and transient TimerService implementations.
+* Implement a distributed implementation of the TimerService SPI.
+** Add a org.wildfly.clustering.ejb.timer SPI to the wildfly-clustering-ejb-spi module to facilitate multiple distributed timer manager implementations.
+** Add a distributed TimerService implementation to the EJB subsystem that delegates to the wildfly-clustering-ejb-spi for timer management and scheduling.
+** Add an Infinispan implementation of the org.wildfly.clustering.ejb.timer API that assigns execution ownership of timers to cluster members based on consistent hashing of the timer identifier.
+*** Leverage the existing Scheduler implementation within wildfly-clustering-ee-cache for timer scheduling
+*** Leverage the existing PrimaryOwnerScheduler decorator that uses consistent hashing of the timer ID to distribute timer execution among cluster members
+* Add a management resource and capability to the distributable-ejb subsystem to configure the Infinispan TimerManagement implementation.
+* Add an persistent-timer-management attribute to the EJB subsystem timer-service resource (as a alternate to the timer-store attribute) that references a TimerManagementProvider capability from the distributable-ejb subsystem.
+** Modify the TimerService DUP to use a CompositeTimerService in the event that the server is configured with a persistent-timer-management, that delegates to the existing TimerServiceImpl for use with transient timers and the DistributedTimerService for use with persistent timers.
+
+=== Nice-to-Have Requirements
+
+* Add an optional transient-timer-management attribute to the EJB subsystem timer-service resource
+** This would replace the existing TimerService implementation completely, by using a distributed timer service that uses a local, non-persistent cache.
+** This has the advantage of supporting passivation of timer instances to support applications with a sufficiently large number of transient timers in excess of some configured maximum.
```

Comment:
No, it is a "nice-to-have requirement".

Reply:
Ok thanks, but will it be a new feature in case the _nice to have_ requirement is implemented?

Reply:
No - this is all one feature.
_Requirements_ represent the scope of the feature.
_Hard requirements_ are requirements essential to the feature (think "MVP")
_Nice-to-have requirements_ are requirements of the feature that should or could be implemented if possible, but are non-essential.
This particular requirement has a high value add, and a low level of effort, thus will almost certainly be included.

Reply:
Thanks for telling me about requirements, which I am aware of.

Now, to clarify my point: since the nice-to-have we're discussing will be implemented, then there will be a new feature, or new behavior, or whatever we want to call it, and it is about passivation of transient timers, right? 
If the behavior doesn't change, then you're right to resolve this thread. If it does change, then the topic should be discussed further.
 
Specifically In such a case, it has to be clarified explicitly and documented so that no functionality is hidden anywhere.